Гринь Надежда

air-breathing opened the pull request.

air-breathing committed:
Update lego.js

maxvipon commented on lego.js:
```diff
@@ -1,8 +1,17 @@
+    args.splice(0, 1);
+    console.log(args);
+    var ind;
+    var collectionA = collection;
```
Этого делать не нужно, потому что в js присваивание переменной объектов (не примитивов) – это не копирование, а всего лишь создание ссылки на этот объект.

maxvipon commented on lego.js:
```diff
@@ -25,3 +34,122 @@ module.exports.limit = function (n) {
+module.exports.sortBy = function(fieldForOrder, order) {
+
+    function sortedRule(a, b){
+        return a[fieldForOrder] - b[fieldForOrder];
```
Для строк это не работает

maxvipon commented:
Первый запуск программы:

``` bash
javascript-tasks-3 git/air-breathing
❯ node index.js
[ [Function], [Function], undefined, [Function], [Function] ]
/Users/maxpon/tmp/urfu/javascript-tasks-3/lego.js:12
        collectionA = args[ind](collectionA);
                               ^

TypeError: args[ind] is not a function
    at Object.module.exports.query (/Users/maxpon/tmp/urfu/javascript-tasks-3/lego.js:12:32)
    at Object.<anonymous> (/Users/maxpon/tmp/urfu/javascript-tasks-3/index.js:10:19)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:475:10)
    at startup (node.js:118:18)
    at node.js:952:3
```

air-breathing committed:
Update lego.js

air-breathing commented:
@maxvipon исправила. Поздно конечно. Если снизите баллы - пойму.

maxvipon commented:
:beetle: раз

запрос:

``` js
lego.query(
    phoneBook,
    lego.select('name', 'birthday'),
    lego.limit(2)
);
```

результат:

``` bash
javascript-tasks-3 git/air-breathing*
❯ node index.js
[ { name: 'Галактион', birthday: undefined },
  { name: 'Анна', birthday: undefined } ]
```

ожидаемый результат:

``` bash
[ { name: 'Галактион' }, { name: 'Анна' } ]
```

maxvipon commented:
:beetle: два

запрос:

``` js
lego.query(
    phoneBook,
    lego.filterIn('favoriteFruit', []),
    lego.limit(2)
);
```

результат:

``` bash
javascript-tasks-3 git/air-breathing*
❯ node index.js
[ undefined, undefined ]
```

ожидаемый результат:

``` bash
[]
```

maxvipon commented:
:beetle: три

запрос:

``` js
lego.query(
    phoneBook,
    lego.sortBy('age', 'desc'),
    lego.limit(2)
);
```

результат:

``` bash
javascript-tasks-3 git/air-breathing*
❯ node index.js
/Users/maxpon/tmp/urfu/javascript-tasks-3/lego.js:10
        collection = args[ind](collection);
                              ^

TypeError: args[ind] is not a function
    at Object.module.exports.query (/Users/maxpon/tmp/urfu/javascript-tasks-3/lego.js:10:31)
    at Object.<anonymous> (/Users/maxpon/tmp/urfu/javascript-tasks-3/index.js:10:10)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:475:10)
    at startup (node.js:118:18)
    at node.js:952:3
```

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+    var arg = [].slice.call(arguments);
+    return function(collection){
+        var first = arg[0];
+        var second = arg[1];
```
В данной реализации функция не будет работать с тремя и более аргументами

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+    var arg = [].slice.call(arguments);
+    return function(collection){
+        var first = arg[0](collection);
+        var second = arg[1](collection);
```
В данной реализации функция не будет работать с тремя и более аргументами

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+            return changedCollection;
+        }
+    }
+    if (order === 'desk'){
```
`desc`

air-breathing replied:
третья ошибка, по ходу, из-за этого и вылетала. починю

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+
+module.exports.filterEqual = function(){
+    var nameField = arguments[0];
+    var valueOfField = arguments[1];
```
вместо этого нужно сразу писать `filterEqual = function(name, value) { ... }`

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+};
+
+module.exports.sortBy = function(fieldForOrder, order) {
+    function sortedRule(a, b){
```
_Tip_: более короткий/читаемый вариант записи:

``` js
function sort(a, b) {
  a = a[fieldName];
  b = b[fieldName];
  return a > b ? 1 : a < b ? -1 : 0;
}
```

maxvipon replied:
А ещё можно сделать фабрику:

``` js
function sort(order) {
  return function(a, b) {
    a = a[fieldName];
    b = b[fieldName];
    switch (order) {
      case 'asc':
        return a > b ? 1 : a < b ? -1 : 0;
      case 'desc':
        return a < b ? 1 : a > b ? -1 : 0;  
    }
  }
}
```

и использовать так

``` js
exports.sortBy = function(fieldName, order) {
  return function(collection) {
    return collection.sort(sort(order))
  }
}
```

maxvipon commented on lego.js:
```diff
@@ -15,13 +24,171 @@ module.exports.reverse = function () {
+
+module.exports.filterIn = function(){
+    var nameField = arguments[0];
+    var valuesOfField = arguments[1];
```
вместо этого нужно сразу писать `filterIn = function(name, values) { ... }`

air-breathing committed:
cloner, sort, or, and 

Добавлено глубокое копирование: честно взятое с Хабра.(вообще в принципе можно было этого и не делать).
Сортировку сделала, как вы рекомендовали.
Маленькие недочеты тоже исправлены.
Так же переделаны or и and на произвольное количество аргументов.

maxvipon commented:
:ok:

maxvipon commented:
:star:
